### PR TITLE
[WIP] Add error message when the backend is not started

### LINKED ISSUE #8565: Improve message when the backend is not started
##### **Overview of the feature request**

When the backend is not started, and you started only the front with `npm start`, try to log into the application.
You will see:

backend-not-started

The goal would be to replace this message, instead of this default.
This message should not change if the credentials are not correct, of course

##### **Related issues or PR**

Similar to this ticket https://github.com/jhipster/generator-jhipster/issues/8564

I opened this ticket, following #8558 (comment) from @jdubois : I'll let add the bounty if you want.

-   [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by fleboulch**
Fix #8565 

My solution relies on HTTP code status. I'm not sure if this approach is really good.

Left TODO
-   [x] Improve timeout customization for jhi-alert-error component (@Input)
-   [ ] Improve multi lines customization for jhi-alert-error component (@Input)
-   [ ] Add translations for all languages
-   [ ] Fix e2e tests

Screenshot (only `npm|yarn start` is running)
not-started-crop

-   Please make sure the below checklist is followed for Pull Requests.

-   [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [ ] Tests are added where necessary
-   [ ] Documentation is added/updated where necessary
-   [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by deepu105** [comment]
@jdubois is this what you wanted? the message needs improvement

File: generators/client/templates/angular/src/main/webapp/app/shared/alert/alert-error.component.ts.ejs [hunk 1]
```diff
@@ -118,33 +138,32 @@ export class <%=jhiPrefixCapitalized%>AlertErrorComponent implements OnDestroy {
...
-                    scoped: true
-                },
-                this.alerts
-            )
-        );
+        const newAlertMessage = newAlert.msg;
         <%_ } _%>
+
+        if (!alertsMessage.includes(newAlertMessage) || newAlert.timeout !== 0) {
+            // this.alerts.push(this.alertService.addAlert(newAlert, this.alerts));
```
Comment (deepu105) [inline generators/client/templates/angular/src/main/webapp/app/shared/alert/alert-error.component.ts.ejs]: we cant have these in generated code, please cleanup

File: generators/client/templates/angular/src/main/webapp/app/shared/login/login.component.html.ejs [hunk 2]
```diff
@@ -24,9 +24,7 @@
 <div class="modal-body">
     <div class="row justify-content-center">
         <div class="col-md-8">
-            <div class="alert alert-danger" *ngIf="authenticationError" jhiTranslate="login.messages.error.authentication">
-                <strong>Failed to sign in!</strong> Please check your credentials and try again.
-            </div>
+            <jhi-alert-error [timeout]=0></jhi-alert-error>
```
Comment (deepu105) [inline generators/client/templates/angular/src/main/webapp/app/shared/login/login.component.html.ejs]: why is timeout 0?

**Comment by fleboulch** [comment]
This PR can't be merged right now as there is still some issues that need to be fixed.

The main goal of this PR was to know if we can rely on http code 504 to determine when the backend is not started. WDYT? 

I need also your opinion about some topics I started to implement here:
# I would like to replace custom `<div class="alert alert-danger"... >... </div>` with the `<jhi-alert-error>` component everywhere. Do you agree with that?
# `<jhi-alert-error>` component should be more flexible if you agree the 1st point. Currently, on the login modal the alert message appears only once (instead of every time in the component) and the message doesn't disappear (instead of a timeout after 5000ms in the component)
#`jhiTranslate` directive needs to be enhanced to help e2e tests because `<span jhiTranslate="login.messages.error.authentication">...</span>` and `const myVar ='login.messages.error.authentication' ; <span jhiTranslate={{myVar}}>...</span> ` doesn't produce same HTML output

According to your feedbacks I will rework on this

**Comment by fleboulch** [comment]
This PR can't be merged right now as there is still some issues that need to be fixed.

The main goal of this PR was to know if we can rely on http code 504 to determine when the backend is not started. WDYT? 

I need also your opinion about some topics I started to implement here:
# I would like to replace custom `<div class="alert alert-danger"... >... </div>` with the `<jhi-alert-error>` component everywhere. Do you agree with that?
# `<jhi-alert-error>` component should be more flexible if you agree the 1st point. Currently, on the login modal the alert message appears only once (instead of every time in the component) and the message doesn't disappear (instead of a timeout after 5000ms in the component)
#`jhiTranslate` directive needs to be enhanced to help e2e tests because `<span jhiTranslate="login.messages.error.authentication">...</span>` and `const myVar ='login.messages.error.authentication' ; <span jhiTranslate={{myVar}}>...</span> ` doesn't produce same HTML output

According to your feedbacks I will rework on this

**Comment by fleboulch** [comment]
This PR can't be merged right now as there is still some issues that need to be fixed.

The main goal of this PR was to know if we can rely on http code 504 to determine when the backend is not started. WDYT? 

I need also your opinion about some topics I started to implement here:
- I would like to replace custom `<div class="alert alert-danger"... >... </div>` with the `<jhi-alert-error>` component everywhere. Do you agree with that?
- `<jhi-alert-error>` component should be more flexible if you agree the 1st point. Currently, on the login modal the alert message appears only once (instead of every time in the component) and the message doesn't disappear (instead of a timeout after 5000ms in the component)
- `jhiTranslate` directive needs to be enhanced to help e2e tests because `<span jhiTranslate="login.messages.error.authentication">...</span>` and `const myVar ='login.messages.error.authentication' ; <span jhiTranslate={{myVar}}>...</span> ` doesn't produce same HTML output

According to your feedbacks I will rework on this

**Comment by fleboulch** [comment]
This PR can't be merged right now as there is still some issues that need to be fixed.

The main goal of this PR was to know if we can rely on http code 504 to determine if the backend is not started. WDYT? 

I need also your opinion about some topics I started to implement here:

1. I would like to replace custom `<div class="alert alert-danger"... >... </div>` with the `<jhi-alert-error>` component everywhere.
2. `<jhi-alert-error>` component should be more flexible. Currently, on the login modal the alert message appears only once (instead of every time in the component) and the message doesn't disappear (instead of a timeout after 5000ms in the component)
3. `jhiTranslate` directive needs to be enhanced to help e2e tests because `<span jhiTranslate="login.messages.error.authentication">...</span>` and `const myVar ='login.messages.error.authentication'; <span jhiTranslate={{myVar}}>...</span> ` doesn't produce same HTML output

According to your feedbacks I will rework on this

**Comment by fleboulch** [comment]
This PR can't be merged right now as there is still some issues that need to be fixed.

The main goal of this PR was to know if we can rely on http code 504 to determine if the backend is not started. WDYT? 

I need also your opinion about some topics I started to implement here:

1. I would like to replace custom `<div class="alert alert-danger"... >... </div>` with the `<jhi-alert-error>` component everywhere.
2. `<jhi-alert-error>` component should be more flexible. Currently, on the login modal the alert message appears only once (instead of every time in the component) and the message doesn't disappear (instead of a timeout after 5000ms in the component)
3. `jhiTranslate` directive needs to be enhanced to help e2e tests because `<span jhiTranslate="login.messages.error.authentication">...</span>` and `const myVar ='login.messages.error.authentication'; <span jhiTranslate={{myVar}}>...</span> ` doesn't produce same HTML output

According to your feedbacks I will rework on this

**Comment by fleboulch** [comment]
This PR can't be merged right now as there is still some issues that need to be fixed.

The main goal of this PR was to know if we can rely on http code 504 to determine if the backend is not started. WDYT? 

I need also your opinion about some topics I started to implement here:

1. I would like to replace custom `<div class="alert alert-danger"... >... </div>` with the `<jhi-alert-error>` component everywhere.
2. `<jhi-alert-error>` component should be more flexible. Currently, on the login modal the alert message appears only once (instead of every time in the component) and the message doesn't disappear (instead of a timeout after 5000ms in the component)
3. `jhiTranslate` directive needs to be enhanced to help e2e tests because `<span jhiTranslate="login.messages.error.authentication">...</span>` and `const myVar ='login.messages.error.authentication'; <span jhiTranslate={{myVar}}>...</span> ` doesn't produce same HTML output

According to your feedbacks I will rework on this

File: generators/client/templates/angular/src/main/webapp/app/shared/login/login.component.html.ejs (hunk 2 above)
Comment (fleboulch) [reply generators/client/templates/angular/src/main/webapp/app/shared/login/login.component.html.ejs]: Currently, in the login modal the alert message doesn't desappear so I keep the same behaviour by adding timeout = 0.

File: generators/client/templates/angular/src/main/webapp/app/shared/alert/alert-error.component.ts.ejs (hunk 1 above)
Comment (fleboulch) [reply generators/client/templates/angular/src/main/webapp/app/shared/alert/alert-error.component.ts.ejs]: Sure! My PR is still WIP and my changes will need to update the ng-jhipster project. That was just to check if tests passed. I'm waiting for some feedbacks to know if I continue or not

File: generators/client/templates/angular/src/test/javascript/e2e/account/account.spec.ts.ejs [hunk 3]
```diff
@@ -51,18 +51,21 @@ describe('account', () => {
         expect(value1).to.eq(expect1);
     <%_ if (authenticationType !== 'oauth2') { _%>
         signInPage = await navBarPage.getSignInPage();
-        await signInPage.autoSignInUsing('admin', 'foo');
+        await signInPage.autoSignInUsing('admin', 'foo bar');
 
         <%_ if (enableTranslation) { _%>
         const expect2 = 'login.messages.error.authentication';
+        const value2 = await element(by.css('.alert-danger > span')).getAttribute('ng-reflect-jhi-translate'); // trim to 30 characters
```
Comment (fleboulch) [inline generators/client/templates/angular/src/test/javascript/e2e/account/account.spec.ts.ejs]: Wrong! I need to tweak <jhi-alert-error>

**Comment by fleboulch** [comment]
@jdubois @deepu105 
I stopped working on this PR because I'm waiting for some feedbacks.
Can we discuss the points I mentioned above?

**Comment by jdubois** [comment]
Sorry @fleboulch I don't have enough time at the moment - I will have a look in a couple of weeks, after Devoxx Belgium

**Comment by jdubois** [comment]
BTW I had a quick look:

- The approach is good
- Of course we will need the same for React & VueJS, but that will be in a second time
- I'm not sure about the HTTP codes, I need to check. And I couldn't see where you checked if we are in dev mode or not, could you give me a pointer?

**Comment by fleboulch** [comment]
Currently I don't check if dev mode is enabled or not.
I will add this check using https://stackoverflow.com/questions/47072426/angular-5-provide-http-interceptor-based-on-environment

**Comment by fleboulch** [comment]
I'm closing this PR as it covers several different topics. I will create separate PRs and try to find a way to fix e2e tests

**Pull request closed without merging**